File: iOS/FaceDetector/FaceDetector/Utilities/AndrewsAlgorithm.swift
//
//  AndrewsAlgorithm.swift
//  FaceDetector
//
//  Created by Sevak Tadevosyan on 25.05.24.
//

import Foundation

import Foundation

public func cross(_ o: CGPoint, _ a: CGPoint, _ b: CGPoint) -> CGFloat {
    let lhs = (a.x - o.x) * (b.y - o.y)
    let rhs = (a.y - o.y) * (b.x - o.x)
    return lhs - rhs
}

public func convexHull<Source>(_ points: Source) -> [CGPoint]
            where Source : Collection, Source.Element == CGPoint {
    
    guard !points.isEmpty else { return Array(points) }

    var lower = [CGPoint]()
    var upper = [CGPoint]()

    let points = points.sorted { a, b in
        a.x < b.x || a.x == b.x && a.y < b.y
    }

    for point in points {
        while lower.count >= 2 {
            let a = lower[lower.count - 2]
            let b = lower[lower.count - 1]
            if cross(a, b, point) > 0 { break }
            lower.removeLast()
        }
        lower.append(point)
    }

    for point in points.lazy.reversed() {
        while upper.count >= 2 {
            let a = upper[upper.count - 2]
            let b = upper[upper.count - 1]
            if cross(a, b, point) > 0 { break }
            upper.removeLast()
        }
        upper.append(point)
    }

    lower.removeLast()
    upper.removeLast()

    return lower + upper
}

File: Android/FaceDetectorAndroid/app/src/main/java/com/example/facedetector/utilities/AndrewsAlgorithm.kt
package com.example.facedetector.utilities

import android.graphics.PointF
import androidx.camera.video.AudioSpec.Source

fun cross(o: PointF, a: PointF, b: PointF): Float {
    val lhs = (a.x - o.x) * (b.y - o.y)
    val rhs = (a.y - o.y) * (b.x - o.x)
    return lhs - rhs
}

fun <Source> convexHull(points: Source): List<PointF> where Source: Collection<PointF> {
    if (points.isEmpty()) { return emptyList() }
    val lower = mutableListOf<PointF>()
    val upper = mutableListOf<PointF>()

    val sortedPoints = points.sortedWith(compareBy({ it.x }, { it.y }))

    sortedPoints.forEach { point ->
        while (lower.count() >= 2) {
            val a = lower[lower.count() - 2]
            val b = lower[lower.count() - 1]
            if (cross(a, b, point) > 0) { break }
            lower.removeLast()
        }
        lower.add(point)
    }

    sortedPoints.reversed().forEach { point ->
        while (upper.count() >= 2) {
            val a = upper[upper.count() - 2]
            val b = upper[upper.count() - 1]
            if (cross(a, b, point) > 0) { break }
            upper.removeLast()
        }
        upper.add(point)
    }

    lower.removeLast()
    upper.removeLast()
    return lower + upper
}